src/Forms/FileUploadable.php

```diff
@@ -0,0 +1,399 @@
+<?php
+
+namespace SilverStripe\Forms;
+
+use Exception;
+use InvalidArgumentException;
+use SilverStripe\Assets\File;
+use SilverStripe\Assets\Storage\AssetContainer;
+use SilverStripe\Core\Object;
+use SilverStripe\ORM\ArrayList;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\DataObjectInterface;
+use SilverStripe\ORM\RelationList;
+use SilverStripe\ORM\SS_List;
+use SilverStripe\ORM\UnsavedRelationList;
+use SilverStripe\ORM\ValidationException;
+
+/**
+ * Provides operations for reading and writing uploaded files to/from
+ * {@see File} dataobject instances.
+ * Allows writing to a parent record with the following relation types:
+ *   - has_one
+ *   - has_many
+ *   - many_many
+ * Additionally supports writing directly to the File table not attached
+ * to any parent record.
+ *
+ * @mixin FormField
+ */
+trait FileUploadable {
+	use Uploadable;
```

Comment:
Having a separate "trait inheritance tree" doesn't really aid readability, but I can't think of an easy alternative (other than rewriting upload behaviour to use composition and a separate class, which would be quite intrusive)